zebra/kernel_netlink.c

```diff
@@ -970,23 +975,25 @@ void kernel_init(struct zebra_ns *zns)
 #endif
 
 	/* Register kernel socket. */
-	if (zns->netlink.sock > 0) {
-		/* Only want non-blocking on the netlink event socket */
-		if (fcntl(zns->netlink.sock, F_SETFL, O_NONBLOCK) < 0)
-			zlog_err("Can't set %s socket flags: %s",
-				 zns->netlink.name, safe_strerror(errno));
-
-		/* Set receive buffer size if it's set from command line */
-		if (nl_rcvbufsize)
-			netlink_recvbuf(&zns->netlink, nl_rcvbufsize);
-
-		netlink_install_filter(zns->netlink.sock,
-				       zns->netlink_cmd.snl.nl_pid);
-		zns->t_netlink = NULL;
-
-		thread_add_read(zebrad.master, kernel_read, zns,
-				zns->netlink.sock, &zns->t_netlink);
-	}
+	/* Only want non-blocking on the netlink event socket */
+	if (fcntl(zns->netlink.sock, F_SETFL, O_NONBLOCK) < 0)
+		zlog_err("Can't set %s socket error: %s(%d)",
+			 zns->netlink.name, safe_strerror(errno), errno);
+
+	if (fcntl(zns->netlink_cmd.sock, F_SETFL, O_NONBLOCK) < 0)
+		zlog_err("Can't set %s socket error: %s(%d)",
```

Comment:
Now that the `netlink-cmd` socket is non-blocking, wouldn't we need to handle EAGAIN errors when sending data to the kernel? This involves scheduling an event to write the remaining bytes once poll(2) tells us that the fd is write-ready. This complexity is the price to pay for non-blocking sockets. I'm not sure if the kernel can return EAGAIN for netlink write operations, but I believe it can if the system is under heavy load.

Reply:
@Jafaral Your reference is for reading from the socket, not writing to it. For readers it just means "nothing to read". I am not aware of any kernel patch that returns EAGAIN on sendmsg for netlink sockets.

Reply:
@dsahern Thanks for your input David. In this case setting `O_NONBLOCK` should be safe here. I just wish this was documented in the netlink(7) man page... this information doesn't seem to be available anywhere.